### DISCUSSION TITLE
Support blogging-like functionality

### DISCUSSION THREAD

**Discussion by fgregg** [observablehq/framework#1199] [Ideas]
I love where this project is going!

I wonder if y'all would be open to considering adding support for [jekyll's blogging pattern](https://jekyllrb.com/docs/posts/).

Namely, that if your markdown file is in a _posts directory, and if its titled like `YEAR-MONTH-DAY-title.MARKUP` then bloggy things happen:

For jekyll, this means that,

1.  The built html is put into a directory structure to get a nice url that looks like DOMAIN/YEAR/MONTH/DAY/title.html
2. the page's date is made available to the layout template, so that the published date can be automatically inserted
3. the site's posts are made available as an object you could interact with `sites.posts` i.e. to show the latest blog post on the front page jekyll lets you do something like this in an `index.md`.

```md
---
layout: default
title: My Cool Blog
---

<h1>{{ site.posts.first.title }}</h1>
<p class="text-muted">{{ site.posts.first.date | date: '%B %d, %Y' }} | <a href="{{ site.posts.first.url }}">Permalink</a></p>

{{ site.posts.first.content }}
```
4. jekyll does other things as well, but those would be the core of it.

I know that blogs are not your first target use case, but I think it could be a great place to go! 

I'd be happy to contribute to this.

**Comment by Fil** [comment]
bloggy things happen? :)

**Comment by fgregg** [comment]
@Fil, sorry for not being more concrete. I put some details in it.

**Comment by Fil** [comment]
Interesting.

I don't know how better it is to write 2024-04-01-april-fool.md than having files in folders such as 2024/04/01-april-fool.md. It's probably convenient for some, but is the gain worth the cost, that is, losing the univocal mapping between file and url, and making relative links even more of a headache? Conversely, if you want a flat folder, can't you live with flat urls?

The "page's date is made available to the layout template" is related to #1036. I do want to have meta data (dates, authors, tags…) available to the page and to server-side rendering (SSR, #413).

Inferring a date from a file name seems like a minor affordance; it could be an application of a hook for inferring the page's meta data from front matter & file contents (it's already the case with the title, which can come from either front matter or from the first H1).

For page templates there was a previous attempt (in #253) with discussions and suggestions that I think are still relevant.

It would be interesting to be able to show a (structured, formatted…) list of all pages on a given page, maybe also through SSR — the SSR function could itself have access to the configuration, or even read all the contents on disk (like minisearch does).

**Comment by Fil** [comment]
I think we should move this to a discussion, and create more actionable feature requests.

**Comment by fgregg** [comment]
great idea, closing this in favor of https://github.com/observablehq/framework/discussions/1198

**Comment by fgregg** [comment]
i think the only reason to prefer the jekyll way of handling dates and filenames is that it would make it easier for people who already have jekyll blogs to adopt framework for their data-intensive blogging. jekyll is, by far, the most popular markdown-based blogging tool, as it's built into github pages.

**Reply by Fil** [reply]
Do you think a script that reads a jekyll repo and "converts" the filenames to framework meta data would help? It could be quick if dates/filenames are the only thing that needs to be adapted. I'm wondering how relative links work, though.

**Reply by altosaar** [reply]
Yes this would work! Claude or GPT-4 should be able to do it; I just started converting my Jekyll website to framework…

**Comment by fgregg** [comment]
> It would be interesting to be able to show a (structured, formatted…) list of all pages on a given page, maybe also through SSR — the SSR function could itself have access to the configuration, or even read all the contents on disk (like minisearch does).

this "logging" feature is probably the minimal necessary and sufficient requirement to support a blog.

**Comment by fgregg** [comment]
are you thinking of a script that takes a file named
`2024-04-01-april-fools.md` with a frontmatter like:

```
===
title: awesome post
===
```

to a markdown file named `april-fools.md` with a frontmatter like:

```
===
title: awesome post
date: 2024-04-01
===
```

?

if, so yes. but the bigger issue is not the input format but what the
output looks like like.

if before my post was at example.com/2024/04/01/april-fools.html and now
it’s at example.com/2024-04-01-april-fools.html or
example.com/april-fools.html then that’s going to break links.

**Reply by Fil** [reply]
If the current URL is example.com/2024/04/01/april-fools.html, the script would have to create docs/2024/04/01/april-fools.md with the date in front matter

**Reply by fgregg** [reply]
got it! yeah, that could be good. I then wonder how you would then collect the files that should go into the list of posts, but you could do that by looking for some attribute in the front matter.

**Reply by altosaar** [reply]
This should already be done in the side bar no?

**Comment by jamesking** [comment]
Long time Jekyll user. I would be very interested in using Framework for blogging and have had the same thought as @fgregg which lead me to this discussion. 

I don't necessarily think that Framework needs to copy the same conventions as Jekyll. 

From my initial investigation, the elements that are missing from Framework that could allow for the creation of a DataVis blog are: 

- A way of marking a subset of the files in the `docs/` folder as "posts" to be included in the blog
- A way of listing these posts chronologically to enable the templating of a blog front page 
- A way to generate and publish an RSS feed from this list of posts

These elements may well be possible now within the feature-set of Framework today. If anyone has a suggestions for implementing them then I would really appreciate pointers.

**Reply by Fil** [reply]
With [page loaders](https://observablehq.com/framework/page-loaders) (released yesterday 🚀), you now have server-side rendering _in any language_. Writing a script that creates an index page listing the posts is a no-brainer. You could even create an index for each tag if you wanted to add tags, with [parametrized pages](https://observablehq.com/framework/params).

The question of marking pages as "posts" is then completely arbitrary (it could be a folder, or all file names that look like a date `2024-09-xxx`, or all files that have "blog" in their front-matter).

This should be enough to get you started. For an early example (which is not a template to follow, just an experiment), see the [D3 gallery](https://observablehq.observablehq.cloud/pangea/d3/) page in pangea ([source code](https://github.com/Fil/pangea/blob/main/src/d3/index.md.ts)). (Although note that atm this one uses the search index to get the list of pages with their titles — but reading the files on disk and their front-matter is not more complicated.)

We don't have a good story for the RSS feed yet, since it wants a fixed URL. But see https://github.com/observablehq/framework/issues/168

**Reply by jamesking** [reply]
@Fil thank you!

**Reply by jamesking** [reply]
I can see how you have constructed an index page in those examples. 

I have a followup question: It is common in other SSG frameworks like Jekyll and Hugo to automatically convert a folder of posts like this…

```
.
├─ src
│  ├─ index.md
│  └─ posts
│      └─ 2024-03-05-post-title.md
│      └─ 2024-03-07-another-post-title.md
│      └─ … etc
└─ ⋯
```

… into an output like this…

```
.
├── dist
│   ├── index.html
│   └── posts
│       └── 2024
│           └── 03
│               ├── 05
│               │   └── post-title.html
│               ├── 07
│               │   └── another-post-title.html
│               └── … etc
└── ⋯
```

I can't work out what the Observable Framework way of doing something similar would be. Is there a way to do something similar that is compatible with Observable's file-based routing?

**Reply by Fil** [reply]
The new [parametrized page loaders](https://observablehq.com/framework/page-loaders) certainly make that _possible_. To make it work we would have to write a script that lists the pages (both for building and for the sidebar), and a page loader that reads them and converts the relative links to scripts and files. Not unsurmountable, but a bit of effort.

**Reply by jamesking** [reply]
Thanks - I actually think it might be better for me to simply use a nested structure in the `src` folder (`./2024/09/19/article-title.md`) rather than go to the trouble of doing something custom for the sake of mimicking Jekyll.

**Reply by jamesking** [reply]
Update on progress:

Here is a "Jekyll-like" pattern that is working nicely for me so far. 

In my `src` folder I keep each blog post as an `.md` file in a date-formatted nested subdirectory like this: `./src/2024/09/19/article-title.md`

I have a helper script [`.helpers/load-posts.js`](https://gist.github.com/jamesking/86a99d0da16f5be1807d149519933a35#file-load-posts-js
) that indexes these posts so that I can `import` them into a [`posts.md.js`](https://gist.github.com/jamesking/86a99d0da16f5be1807d149519933a35#file-posts-md-js) page-loader and display an index of posts at the `/posts` URL. This uses the same `gray-matter` dependency that Observable Framework already uses, so I'm not adding any more dependencies. 

I also realised that I could import this list of posts into my [`observable.config.js`](https://gist.github.com/jamesking/86a99d0da16f5be1807d149519933a35#file-observablehq-config-js) file so that I can populate the `pages` configuration automatically without needing to keep the list of posts updated.

Side note: It is really neat that Framework's config is a `js` file as opposed to `json` or `yaml` so tricks like this are possible.

My code is in this [gist](https://gist.github.com/jamesking/86a99d0da16f5be1807d149519933a35) in case it helps others.

**Reply by mcmcclur** [reply]
> I also realised that I could import this list of posts into my [`observable.config.js`](https://gist.github.com/jamesking/86a99d0da16f5be1807d149519933a35#file-observablehq-config-js) file so that I can populate the `pages` configuration automatically without needing to keep the list of posts updated.

That's pretty much what I do here: https://mcmcclur.observablehq.cloud/mark-on-framework/

The list of dated posts is generated automatically from the config file. A key difference, though, is that I simply copy the config file over into my src directory so that Javascript in the pages can read it. I was literally just wondering if page loaders have access to path information.

Also, I think it might be worth mentioning that the Framework project linked above embeds GitHub Discussions via Giscus as the commenting system.

Thanks!

**Reply by Fil** [reply]
thanks for sharing! I think you could simplify the recursive scanning by using `import {globSync} from "glob";` ([glob](https://www.npmjs.com/package/glob) is already a dependency of framework).

**Comment by Fil** [comment]
I've finally managed to support opengraph and twitter cards (#178) in pangea; see for example [this page](https://www.opengraph.xyz/url/https%3A%2F%2Fobservablehq.observablehq.cloud%2Fpangea%2Fplot%2Fvoronoi-treemap).

Since we need absolute URLs and *since it only makes sense for public projects*, the trick is to use the github repo as the source for the images, instead of adding the images as assets in the data app. 

The code can be seen here: https://github.com/fil/pangea/blob/main/observablehq.config.ts#L97 ; 
([commit](https://github.com/Fil/pangea/commit/0a79fe8c3d03c917fc2dd8b3f9fc1f8af033901a)).

My implementation is incomplete (lacking a few of the usual fields), but they should be pretty simple to add. Not sure how to fill the description… maybe from the front-matter — or we ask claude to summarize the page?

**Reply by Fil** [reply]
Note to self: an alternative technique is to reference the assets as files in Framework, and compute their hash the exact same way framework does. It goes something like this:

```js
import {createHash}  from "node:crypto";

function head({path, title}) {
  const root = "src";
  const ROOT_URL = "https://observablehq.observablehq.cloud/pangea/";
  let image = "";
  try {
    // computes the same hash as framework 🌶
    const contents = readFileSync(join(root, `thumbnail${path}.png`));
    const key = createHash("sha256").update(contents).digest("hex").slice(0, 8);
    const esc_image_url = JSON.stringify(`${ROOT_URL}_file/thumbnail${path}.${key}.png`);
    image = `<link href="/thumbnail${path}.png">
<meta property="og:image" content=${esc_image_url}>
<meta property="twitter:image" content=${esc_image_url}>
`;
  } catch (error) {}

  return `<meta property="og:title" content=${JSON.stringify(
    title
  )}>\n${image}<link rel="icon" href="data:image/svg+xml,<svg xmlns=%22http://www.w3.org/2000/svg%22 viewBox=%220 0 100 100%22><text y=%22.9em%22 font-size=%2290%22>${EMOJI_FAVICON}</text></svg>">
  `;
}
```

This adds a `link[href]` that makes Framework build the asset, and in parallel recomputes the same hash. This approach is a bit complicated, but it could be encapsulated in a specialized function.

**Reply by mcmcclur** [reply]
> I've finally managed to support opengraph and twitter cards (#178) in pangea; see for example [this page](https://www.opengraph.xyz/url/https%3A%2F%2Fobservablehq.observablehq.cloud%2Fpangea%2Fplot%2Fvoronoi-treemap).

I'm not sure if you noticed when [I announced my Framework site on the fourm](https://talk.observablehq.com/t/my-framework-site-with-blog-like-coments/9716) but, if you check out the link, you should recognize it as opengraph enhanced. I set that up by simply using the fact that the `head` on the configuration file can be a function depending on the title of the page. Thus, you can place any `meta` tags that you'd like in there - including opengraph tags. 

Having said that, I definitely agree that tools built specifically to support opengraph and SEO more generally would be very welcome!

**Comment by Fil** [comment]
I wrote a postbuild script to add a simple sitemap [to pangea](https://observablehq.observablehq.cloud/pangea/sitemap.txt), with the proper headers that reference it.

https://github.com/Fil/pangea/blob/e179717808c66526d5e17a933245f7fe6b91851b/package.json#L7-L8

https://github.com/Fil/pangea/blob/e179717808c66526d5e17a933245f7fe6b91851b/observablehq.config.ts#L101

Making a RSS/Atom file could follow the same approach.

**Reply by jamesking** [reply]
I have attempted to create an RSS feed, but I hit a bit of a wall. 

My first idea was to user Framework's page-loaders by creating a file `/src/feed.atom.js` but I realise that page-loaders only emit markdown, so I don't think that route it going to work.

@Fil - I see you are using `jq` and a post-build step to process the `minisearch.json` into a Sitemap.  I guess I could also use a post-build step, but is there any way I could do this during build time instead?

_Thinking out loud: It would be awesome if page-loaders could emit more than just markdown._

**Reply by Fil** [reply]
Currently that's not possible, as loaders can only emit a page, a file, or an embeddable script. Pages are processed as Markdown and into HTML, files are hashed, and embedded analytics are JavaScript.

What's possible _today_ is to build the feed either in a post-build script, or maybe, if that is more convenient, as a file with a very simple post-build step (something like `cp dist/_file/feed.*.xml dist/feed.xml`).

Using non-hashed files has many drawbacks in general, so I'm not sure it's a pattern that we'd want to encourage. Things like atom feed, sitemaps etc look like exceptions to that rule—because of their peculiar design. (There would be not such issue if the Web had adopted [microformats](https://microformats.org/wiki/h-feed), for example!)

**Reply by mbostock** [reply]
I have been considering supporting data loaders that generate stable paths under the embedded analytics feature. It could be used e.g. to create a parameterized route SVG that can be embedded in another application. It could also be used for a feed.atom. It would be the same contents as _file but without the _file and without the content hash. And you’d need to list these routes in dynamicPaths.

**Reply by jamesking** [reply]
@Fil Thanks for the quick response. 

> What's possible today is to build the feed either in a post-build script, or maybe, if that is more convenient, as a file with a very simple post-build step (something like cp dist/_file/feed.*.xml dist/feed.xml).

I am trying this, but right now I am running into a strange [issue](https://github.com/observablehq/framework/discussions/1679) with data loaders in general.

@mbostock Thank you. I think supporting stable paths could be helpful and intuitive for these kinds of situations, but the post-build step that @Fil recommends would be fine for me today (if I can get it working…)

**Reply by jamesking** [reply]
@mbostock @Fil I actually think that some kind of "static asset loader" would be a great addition to the data, page and embeddable script loaders. 

This would make it easy to publish RSS, OG images and other "blogging-like" things. 

More fundamentally, it would enable Framework to publish assets at a fixed URL for others to consume. This is very similar to the ethos behind Datasette which I like very much.

**Reply by jamesking** [reply]
> What's possible today is to build the feed either in a post-build script, or maybe, if that is more convenient, as a file with a very simple post-build step (something like cp dist/_file/feed.*.xml dist/feed.xml).

- [x] I have a data loader `src/feed.atom.js` that generates an atom feed successfully. 
- [x] I ensure that the data loader runs by [referencing it in a page](https://github.com/observablehq/framework/discussions/1679#discussioncomment-10742390)
- [x] I have a post-build script that copies the resulting file this to a fixed location: `cp dist/_file/feed.*.atom dist/feed.atom` and I can see that the file is copied correctly to the `/dist` directory.
- [ ] **But** when I navigate to `/feed.atom` I get a 404 error 

What am I missing?

**Reply by mbostock** [reply]
How are you serving `dist`? Are you deploying to Observable Cloud? The `/feed.atom` URL won’t work in preview. (The preview server doesn’t look in `dist`.)

**Reply by jamesking** [reply]
I see, thanks! I hadn't appreciated that the preview server wasn't loading files from the `dist` folder.

**Reply by jamesking** [reply]
OK, so RSS is _mostly_ [working](https://jamesking.io/feed.atom). 

My last hurdle is related to rendering markdown content of posts into the feed using `markdown-it` ([gist](https://gist.github.com/jamesking/9c06a5196c4c89afe066468282de2b90#file-feed-atom-js-L35)). This was working fine until I added a [post with an image](https://jamesking.io/2024/09/26/expected-value-dice). Because the image file gets hashed I need to know what the hashed file location will be when generating the feed.

Is there a Framework function I can call that will render the contents of a post, but with hashed references?

**Reply by mbostock** [reply]
You could build the site (`observable build`) and then consume the generated HTML.

**Reply by jamesking** [reply]
Nice idea, thanks. I'll give it a go.

**Reply by jamesking** [reply]
Got it working and it is a good solution. Thanks for the suggestion.

**Comment by Fil** [comment]
Adding a script to adopt a `src/public/` folder that gets copied onto the `dist/` folder before deploying is as simple as:

```json
  "predeploy": "cp -r src/public/* dist/",
  "deploy": "yarn predeploy && yarn observable deploy"
```

All of this can be done user-side without too much complexity. All we need now is a template (repo) so that users who don't want to have to think through all of this just have to clone the repo to start their blog.

**Comment by Fil** [comment]
I had seen the announcement and tested the comments system, but hadn't seen
the opengraph tags. Can you share how you obtain the absolute url for
og:image?

**Reply by mcmcclur** [reply]
Well, you know where the file is going to go and you know (mostly) what the name is. For example, if I've got the following file in my source:

    src/samples/images/groovy_pic.png

Then I know I'm going to have 

    https://wncviz.com/_file/samples/images/groovy_pic.xxxxxxxx.png

where `xxxxxxxx` is the hash for the file.  You just need to run

    npm run build

to find out the hash. Since an image file has no dependencies, you can be sure that it won't change until you change the file.

Obviously, that's a bit of a pain and it would be preferable to have a dedicated approach to automate this procedure. I do hope you include such a tool in the future but I think it's worth the trouble to add these meta tags via the `head` function for the time being, if you're trying to build a site that might gain some traction.

**Reply by Fil** [reply]
OK, I thought you had found a clever workaround :)  You have to assume that this file is also statically linked somewhere else in the page, then. Otherwise, you have to add it to the head too, like I did in https://github.com/observablehq/framework/discussions/1199#discussioncomment-10624165.

**Reply by mcmcclur** [reply]
> You have to assume that this file is also statically linked somewhere else in the page

Yes, that's true. At one point, I referenced an image that I didn't want to actually display and then hid it with CSS - just so I could I could have an `og` tag. A bit of a pain but it worked. :)

**Comment by jamesking** [comment]
I thought I would post a summary of where I have got to with implementing a blog using Framework. Here is the (very minimal) [blog](https://jamesking.io) if you would like to take a look. My motivation in choosing Framework is that I eventually want to be able to add posts with interactive explanations and data visualisations (using Observable Plot and D3), and host them on my own domain.

Observable Framework is mainly designed for publishing documentation sites and dashboards, so there are some features missing that other static site generators have (like Jekyll, and Astro) which need to be patched in.

### Creating an index of posts. 

I have previously written how I implemented this here: https://github.com/observablehq/framework/discussions/1199#discussioncomment-10738633.

Issues: 
- In preview, the page that displays an index of posts (a page loader `posts.md.js`) does not update automatically when I add or edit a new post. I currently workaround this by adding and deleting a few whitespace characters in the `posts.md.js` to nudge it to rebuild that page.

### Adding a feed

I created a data loader `feed.atom.js`  that uses the [feed](https://github.com/jpmonette/feed) library to build an atom feed. Code here: https://gist.github.com/jamesking/9c06a5196c4c89afe066468282de2b90

Because data loaders publish a hashed file, I also have to edit the `build` script in `package.json` to `"observable build && cp dist/_file/feed.*.atom dist/feed.atom"`

To ensure that Framework builds my feed I also needed to reference the data loader from a page, so I have a hidden page containing a single line: `FileAttachment("feed.atom")` 

Issues: 
- It is not possible to view/test the feed when running the Observable preview server. Navigating to `/feed.atom` returns a 404 because the preview doesn't load files from the `/dist` folder which the `build` command builds to. 
- I am using `markdown-it` to render a post's `md` content into the feed. Currently this breaks images embedded in the file, as my script does not know that the image URL will be hashed. Problem described in my comment above https://github.com/observablehq/framework/discussions/1199#discussioncomment-10765456. 
- When rendering the content of a post into the feed, I currently have a hack to remove the first `h1` tag in the article body. Otherwise, the title of the post (as defined by the post's frontmatter `title` attribute) will be repeated in the article. Some kind of templating system would be really helpful here. 

### Hosting

I am currently using Netlify to build the site, which "just worked" when I connected the project on Github. I am not sure if this will be as easy if I started to rely on non-js data loaders that require installing or if I start to require network calls to load data when building the site.

### Wishlist

Some kind of templating system would be very useful. Different types of pages could be rendered using different templates. Posts could be rendered with a `post` template that might differ from a `page` or a `gallery` template for example.

A way to render files to static locations for specific reasons seems important. Hashing _all_ files feels like it is causing problems, especially for images that might need to be consumed from clients like RSS readers, or Open Graph previews on social media.

**Reply by jamesking** [reply]
Update on adding a feed: 

Thanks to a [suggestion](https://github.com/observablehq/framework/discussions/1199#discussioncomment-10765874) from @mbostock I ended up converting my `feed.atom.js` data loader into a plain node script, `build-feed.js` ([gist](https://gist.github.com/jamesking/3dbc6bdf10bd9106d60dbd6c140d9176)). This loops through the rendered posts in the `.dist/` folder and builds an RSS feed with the `content` for each post copied directly from the rendered files. 

I run this as a post-build step once the site has been built: `observable build && node ./helpers/build-feed.js` in `package.json`.

**Reply by jamesking** [reply]
Note to self: Here is a useful summary of @Fil's implementation of open graph images.  https://github.com/observablehq/framework/issues/178#issuecomment-2395757913 to work on next.

**Reply by jamesking** [reply]
Just an update, as I have continued to build on Framework and have made numerous small changes since my comment above, all with the help of Claude. I am really pleased with the changes and thought I would share, in case it helps anyone else building on top of Framework.

- I implemented @Fil's [suggested](https://github.com/observablehq/framework/issues/178#issuecomment-2395757913) implementation for hosting Open Graph images and computing a hash separately from Framework's build process. Works really well. I can now put an image next to a page in the `src` directory and reference it in the frontmatter of the page. This will get included in the build and the image's eventual url can be referenced as the page's `og:image` in the `head`. 
- I implemented a lightweight templating system, inspired by Jekyll, so pages for posts and other content can have consistent template content added such as titles, and publication date. This uses a combination of `parameterised-routes` and `page-loaders`. There is also a trick to making this work nicely: Posts are written to a `yyyy/mm/dd/slug.post` file rather than an `.md` file. This prevents Framework from also building a page for each "untemplated" post. I also wrapped `observable preview` with a helper script that to watches for changes to these .post files and then re-writes their associated `page-loader` file. This makes live preview work nicely in development without needing to do a full rebuild.

I know that coding agents are creating a lot of issues for open source-maintainers with a swamp of spammy PRs. On the flip-side, Claude meant I could extend Framework for my own needs, without needing to submit any a PR or make any feature requests. Its been a great foundation to build on.

**Comment by altosaar** [comment]
I am also using observable framework as a blog - appreciate the inspiration @jamesking ! I spent the weekend migrating examples from earlier this year I made to my personal website/portfolio as I transition out of faculty roles and find a position in industry:

https://jaan.li

Open to any feedback! It's barebones, and I still need to add the og:image functionality from https://github.com/observablehq/framework/issues/178#issuecomment-2344135840.

We will get there! Thank you so much for enabling this! I'm migrating https://jaan.io over to jaan.li from jekyll to observable framework now, in case anyone has tips/pointers :)
